### DISCUSSION TITLE
Adapting the Framework to SDL2/OpenGL

### DISCUSSION THREAD

**Discussion by MonadMania** [rabbibotton/clog#347] [Q&A]
I really like CLOG's API. I'm working on a game using Common Lisp's wrappers for SDL2/OpenGL as a starting point. I'd like to make a UI toolkit so that it's easy to change the UI while the game is running. Has anyone adapted this API to a desktop GUI? How does HTML/CSS/JS fit into that, and is it possible to replace replace those tools (for example, I don't see myself writing a JS engine)?

**Comment by rabbibotton** [comment]
The Builder is a good example of a full desktop application in CLOG, if you would like to see it as a native app (no browser) go to the clogframe ide and run make-linux (or make-mac, for Windows see the README or use the https://github.com/rabbibotton/clog-win64-ez version and see README there) then run make-builder and run the builder exectutable.

I would never create a UI today in anything but CLOG and if have no choice html, it simple makes no sense anymore to use any toolkit. However... SDL/OpenGL there is room what I would do is right the entire UI in CLOG and then (since CLOG is a native app just using the browser or html controls for rendering a GUI) I would just pop a second window that your CLOG UI is controlling that uses an OpenGL or SDL context.

If I _had_ to make it all run on one screen... I would use the C++ code from clogframe (as fully cross platform) and modify it to provide the SDL context and Browser control or say GTK again just to house a browser control and an SDL or OpenGL context.

**Reply by MonadMania** [reply]
I look forward to trying this out. Thank you for your answer. I'll link a demo here if I succeed.

**Comment by aykaramba** [comment]
MonadMania, if you get an SDL2/OpenGL context started and figure out how to get webview rendering in a single window (let's say as a transparent layer over top of base GL rendering of something like a HUD), can you post same sample code or just jump in here and provide a hint on how this might be done?

I have read a bunch about webview and there are a lot of people wanting to do this but not too many code samples of how it can be done.

This is an intriguing idea, to use webview and have CLOG provide the interface components while the GL rendering happens in the layers underneath with full 3d acceleration and compositing working.

That is kind of a revolutionary thought in my view.

**Reply by rabbibotton** [reply]
It is not difficult at all to do what you just said. Just use the CLOGframe code then create the webgl or sdl context on top of it with a transparent background

**Reply by aykaramba** [reply]
Oh!  Okay, will dig into that shortly then.  Thanks for the heads up.
